stacks-signer/src/runloop.rs

```diff
@@ -937,21 +1069,98 @@ impl<C: Coordinator> SignerRunLoop<Vec<OperationResult>, RunLoopCommand> for Run
     }
 }
 
+/// When this threshold is reached, the signer will use the NACKed CoordinatorMetadata if that data
+/// is more updated than its own chain view
+fn get_nack_threshold(total_signers: u32) -> u32 {
+    (total_signers + 1) / 2
+}
+
+/// Clean up outdated coordinator_metadata NACKs that have a stale or equivalent chain view compared to the current signer
+fn cleanup_outdated_coordinator_nacks(
+    nack_messages_received: &mut HashMap<CoordinatorMetadata, HashSet<u32>>,
+    current_observed_height: u64,
+) {
+    let outdated_metadata_to_remove: Vec<CoordinatorMetadata> = nack_messages_received
+        .iter()
+        .filter(|(coordinator_metadata, _)| {
+            coordinator_metadata.stacks_block_height <= current_observed_height
+        })
+        .map(|(coordinator_metadata, _)| *coordinator_metadata)
+        .collect();
+
+    for coordinator_metadata in outdated_metadata_to_remove {
+        nack_messages_received.remove(&coordinator_metadata);
+    }
+}
+
+/// Returns the coordinatorMetadata broadcasted by most signers if it meets the NACK threshold
+fn get_lead_coordinator_by_nack(
+    nack_messages_received: &HashMap<CoordinatorMetadata, HashSet<u32>>,
+    nack_threshold: &u32,
+) -> Option<CoordinatorMetadata> {
+    nack_messages_received
+        .iter()
+        .filter(|(_, signers)| signers.len() >= *nack_threshold as usize)
+        .max_by_key(|(_, signers)| signers.len())
+        .map(|(coordinator_metadata, _)| *coordinator_metadata)
+}
+
+/// Create coordinator Info with fake consensus hash and block height to not block signers if they can't calculate
+/// coordinator properly even after retries. This will result to NACKs and will eventually get ignored/overthrown
+fn get_default_coordinator_info(signer_id: u32, public_key: PublicKey) -> CoordinatorInfo {
+    let default_coordinator_metadata = get_default_coordinator_metadata();
+    CoordinatorInfo::new(signer_id, public_key, default_coordinator_metadata)
+}
+
+fn get_default_coordinator_metadata() -> CoordinatorMetadata {
+    CoordinatorMetadata::new(ConsensusHash([0u8; 20]), 0)
+}
+
+/// Attempts to return the most up-to-date CoordinatorMetadata based on the current signer's view
+/// and the NACK messages received. If NACK messages indicate a majority view with a higher block height
+/// than the current signer's, the function updates to that view. Otherwise, it retains the current signer's view
```

Comment:
I don't think this is what we want to do.  The signer should be querying its Stacks node for the latest chain view, and _only_ its Stacks node.  The Stacks nodes take care of making sure they all have the same chain view as one another.  What we don't want is to allow other signers to compel the local signer to assume a chain view that its local node does not believe exists.  If this happened, then the data returned by `stacks_client.rs` would be inconsistent with the signer's chain view, and thus unreliable and useless for making progress on signing rounds and DKG.  For example, the signer would be unable to validate blocks since it would have two conflicting views of the Stacks chain tip.

Reply:
@jcnelson For 'The Stacks nodes take care of making sure they all have the same chain view as one another,' then when would the case arise where one signer might have a more outdated view? Are we suggesting that it's possible, though not a frequent event, since the nodes are already tasked with ensuring consistency? If the NACKs aren't necessary for determining the coordinator, could I remove that logic and simply use the NACKs for logging purposes only? If that's the case, I think I'll need to revise my NackMessage as it currently includes the signer_id of the one that is more updated and broadcasts its CoordinatorMetadata, which isn't directly aimed at the signers who are falling behind.

Reply:
@jcnelson if you could confirm, I can proceed by deleting the nack_threshold part and how its used as the local signer should keep its coordinator calculation as is and only use NACKs for logging purposes and for that I will revise my NackMessage a bit too.

Reply:
> when would the case arise where one signer might have a more outdated view?

It is possible that one signer's node is behind the chain tip.

> Are we suggesting that it's possible, though not a frequent event, since the nodes are already tasked with ensuring consistency?

Correct.

> If the NACKs aren't necessary for determining the coordinator, could I remove that logic and simply use the NACKs for logging purposes only? 

Yes, that is what I'm suggesting.  The local signer should not be tasked with assessing whether or not its view of the chain is fresh vis-a-vis what the remote signers say.  However, the local signer _may_ treat a threshold of NACKs as a _hint_ to back off of their current request for a time, in order to give their local node a chance to catch up (but since this is a hint, the signer may ignore it as a local policy decision or choose what the back-off delay will be).

Reply:
@jcnelson Thank you for clarifying. I pushed the fixes now by removing the NACK logic from coordinator calculation. I updated message types to include the signer_id of the sender and receiver for better logging. For the local policy decision aspect, I expanded the configs with default behavior to be ignoring the hints (can change this if needed). 
The back-off delay is a simple one-time thread sleep, but if needed, I could make this a repeated back-off up to the configured duration during which I query the Node to see if the view has been updated or not.
I'm working on adding more unit tests now, but happy to make other adjustments if you get a chance to look at my changes so far.